Turn BER decoder into a suspendible generator

etingof opened the pull request:
The goal of this change is to make the decoder stopping on input
data starvation and resuming from where it stopped whenever the
caller decides to try again (hopefully making sure that some more
input becomes available).

janpipek committed:
Prepare for streams

Rewrite Decoder in terms of BytesIO

BER Decoders implemented with BytesIO but for the most complex

BER UniversalConstructedTypeDecoder in terms of BytesIO

BER Decoder (stream-based) suggestion

Fixed some of the failing tests

Fixed several failed tests

Fix all remaining tests but the non-implemented Any

Implement untagged Any with back-seek

Fix cer and der to work with streams

Simplify unnecessary added complexity

Make use of IOBase hierarchy (properly?) - in progress

Tests failing

Fixed most failing tests

1 remaining

Severaů small optimizations

Fix logging

Note: As we do not want to read the whole stream, explicit output of remaining bytes is not used.

Rename and document utility functions for BER decoder

Fixed ínverted condition in BitStringDecoder.valueDecoder

Fixed wrongly acquired fullPosition in AnyDecoder.indefLenValueDecoder

Fixed logging None length

endOfStream(BytesIO) working in 2.7

Microoptimizations for endOfStream (not using it)

Test for checking binary files as substrate

Python 2.7 BytesIO wrapper for `file`s

Refactor keep API compatibility with original version

janpipek committed:
API that work with pyasn1-modules

etingof committed:
Improve open maps debugging in decoder

janpipek committed:
Fail with unseekable streams.

janpipek committed:
UnsupportedSubstrateError

janpipek committed:
Update tests with more streams for ber.decoder

janpipek committed:
Trivial changes from the MR.

janpipek committed:
Docstrings in requested format.

janpipek committed:
Implement _CachedStreamWrapper

janpipek committed:
Additional test on ZIP files

etingof committed:
Fix failing `OctetString` to unicode conversion tests on Py2

janpipek committed:
Address several pull requests comments + hide asSeekableStream

janpipek committed:
Hide other auxiliary functions.

janpipek committed:
Simplify _CachingStreamWrapper

janpipek committed:
CachingStreamWrapperTestCase

etingof committed:
Refactor BER decoder into a suspendable coroutine

The goal of this change is to make the decoder stopping on input
data starvation and resuming from where it stopped whenever the
caller decides to try again (hopefully making sure that some more
input becomes available).

This change makes it possible for the decoder to operate on streams
of data (meaning that the entire DER blob might not be immediately
available on input).

On top of that, the decoder yields partially reconstructed ASN.1
object on input starvation making it possible for the caller to
inspect what has been decoded so far and possibly consume partial
ASN.1 data.

All these new feature are natively available through
`StreamingDecoder` class. Previously published API is implemented
as a thin wrapper on top of that ensuring backward compatibility.

etingof committed:
Refactor BER decoder into a suspendable coroutine

The goal of this change is to make the decoder stopping on input
data starvation and resuming from where it stopped whenever the
caller decides to try again (hopefully making sure that some more
input becomes available).

This change makes it possible for the decoder to operate on streams
of data (meaning that the entire DER blob might not be immediately
available on input).

On top of that, the decoder yields partially reconstructed ASN.1
object on input starvation making it possible for the caller to
inspect what has been decoded so far and possibly consume partial
ASN.1 data.

All these new feature are natively available through
`StreamingDecoder` class. Previously published API is implemented
as a thin wrapper on top of that ensuring backward compatibility.

etingof commented:
@janpipek Here's my thoughts on suspendible decoder in form of some code. 

It's still WIP, unit tests won't pass.

On the bright side, any decoder would gracefully handle as little or as much data as it is given allowing the caller to collect some more input and repeat the decoder call resuming it from where it yielded.

Would this be a viable design, WDYT? How does it stand performance-wise?

etingof commented on pyasn1/codec/ber/decoder.py:
```diff
@@ -67,8 +67,18 @@ def read(self, n=-1):
                 return read_from_cache
 
         read_from_raw = self._raw.read(n)
+        if read_from_raw is None:
```
This could only happen in `O_NONBLOCK` mode.

etingof commented on pyasn1/codec/ber/decoder.py:
```diff
@@ -115,7 +125,7 @@ def _asSeekableStream(substrate):
     """
     if isinstance(substrate, bytes):
-        return BytesIO(substrate)
+        return _CachingStreamWrapper(BytesIO(substrate))
```
Should we always wrap input stream? Perhaps, no? The only reason for that here is to pass all reading through `_CachingStreamWrapper.read` generator routine.

etingof commented on pyasn1/codec/ber/decoder.py:
```diff
@@ -276,13 +285,16 @@ def valueDecoder(self, substrate, asn1Spec,
 
-        the_bytes = substrate.read(length)
+        for the_bytes in substrate.read(length):
+            if the_bytes is None:
```
`None` here indicates insufficient input. This loop is kinda stinky...

etingof commented on pyasn1/codec/ber/decoder.py:
```diff
@@ -276,13 +285,16 @@ def valueDecoder(self, substrate, asn1Spec,
+
         if not the_bytes:
-            return self._createComponent(asn1Spec, tagSet, 0, **options)
+            yield self._createComponent(asn1Spec, tagSet, 0, **options)
```
We might have to yield on insufficient read, for that reason we become a coroutine as well. So we have to yield everywhere else and our caller has to treat us as a generator coroutine.

janpipek commented on pyasn1/codec/ber/decoder.py:
```diff
@@ -1447,313 +1631,336 @@ def __call__(self, substrate, asn1Spec=None,
+                            yield
+
+                    if not firstByte:
+                        yield
```
I'd guess this (we read zero bytes) is semantically different from `firstByte` being None (bytes not available). Should we `return` instead?

janpipek replied:
Actually, this cannot happen as a zero-length bytes is never yielded from _read.

janpipek commented on pyasn1/codec/ber/decoder.py:
```diff
@@ -171,15 +177,53 @@ def _peek(substrate, size=-1):
+        yield
+
+    if len(received) < size:
+        substrate.seek(len(received), os.SEEK_CUR)
```
Should this be MINUS len(received)?

janpipek commented:
Thanks!

It looks very elegant. Though using a for cycle to produce exactly one value looks a bit confusing at first sight, it makes perfect sense. Thanks for showing me how the coroutine approach works :-)

I added two random comments that occurred to me when browsing through the code. I will have a more detailed look later and I will make some performance tests. I am a bit sceptical as calling `_read`, `_peek` incur some costs. Let's see though.

If the performance cost is not too high, the improved universality of the solution might still be worth it. In the end, you are the judge in this dilemma.

etingof commented:
Awesome! Just keep in mind that I am still refactoring the PR.

I am thinking to may be yield the ASN.1 object we've managed to build so far along with a flag indicating whether we are done or not... That would let the user to consume and possibly trash partial data as it arrives. That would target the case of infinitely large input.

The other thought I have is that a theoretical alternative to the stream abstraction could be this [.send()](https://docs.python.org/2/reference/expressions.html#generator.send) generator method feeding substrate to the suspended generator thus letting it to proceed some further. Though stream seems also a good fit for the purpose.

janpipek commented:
Performance-wise, it looks like this for small objects:

![image](https://user-images.githubusercontent.com/92804/65028467-61960e80-d93c-11e9-83a5-53f5fbd0f070.png)

I was not able to parse a large file though (yet). It complains about a tag not being present in the spec.

etingof commented:
Yeah, it's slower and not quite functional yet. Let's see how it stands once I iron it out.

etingof commented:
> Though using a for cycle to produce exactly one value looks a bit confusing at first sight, it makes perfect sense. 

Once we drop Python 2.7, `yield from` should replace these silly loops entirely.

janpipek commented:
Just wanted to add some numbers. Looks similar in performance to my previous iteration before I started eliminating unnecessary function calls.

So, I will wait for un-WIP-ping :-)

etingof commented:
> Just wanted to add some numbers. Looks similar in performance to my previous iteration before I started eliminating unnecessary function calls.
> 
> So, I will wait for un-WIP-ping :-)

Thank you! It's important not to depart too much from the performance improvement goal we have  in the first place!

etingof commented:
BTW, it occurred to me that may be it would make sense, for consistency's sake (at least), to refactor BER encoder so that it would produce a stream rather than a string...

Not sure if string concatenation is so much expensive compared to string cutting (apparently!). If it is not, there won't be a performance gain.

However, the actual streaming could happen when the encoder iterates over a `SEQUENCE OF`, which  yields if existing elements are already encoded, but the sequence is not over yet.

etingof commented:
This is not forgotten! o/

I've updated my patch so that unit tests seem to pass.

That's quite a huge patch, sorry about that!

etingof commented:
Oh, this change is getting out of hand side-wise...

On the other hand, it's hopefully better structured now. I plan on some more minor refactoring and documenting things here and there.

The new proposed API looks like this:

```python
stream = io.BytesIO(b'0\t\x02\x01\x01\x02\x01\x02\x02\x01\x03')

schema = SequenceOf(componentType=Integer())

decoder = StreamingDecoder(stream, asn1Spec=schema)
for asn1Object in decoder:
    print(asn1Object)
```

The `stream` should be read from and one or more asn1Object's emitted. If stream data is insufficient, `SubstrateUnderrunError` instance is returned and the caller is supposed to fill in the stream and keep iterating.

janpipek commented on pyasn1/codec/ber/decoder.py:
```diff
@@ -303,24 +156,45 @@ def valueDecoder(self, substrate, asn1Spec,
+        if not length:
+            raise error.PyAsn1Error('Empty BIT STRING substrate')
+
+        for chunk in streaming.isEndOfStream(substrate):
```
Performance nit-pick: I'd import all names explicitly (to be applied in here and in other places) so that you don't have to access the module's namespace.

On my computer with Python 3.7, the extra dot adds 20 ns per call.

janpipek replied:
https://github.com/janpipek/various_benchmarks/blob/master/python/access_penalties.ipynb

etingof commented:
> Is this reuse of instance motivated by performance? Otherwise it looks like a strange pattern.

Yes. If your measurements do not reveal signifiant difference compared to `isinstance` call, we can drop that.

janpipek commented on tests/codec/ber/test_decoder.py:
```diff
@@ -1669,8 +1658,11 @@ def testInvalidFileContent(self):
-                    _ = list(decoder.decodeStream(source))
+                list(decoder.decodeStream(source))
+
+        except EndOfStreamError:
```
Is it possible to use `assertRaises`? If we expect an exception, we should check that it really was raised.

etingof replied:
> Is it possible to use `assertRaises`? If we expect an exception, we should check that it really was raised.

Sure! I feel like we should first convert all legacy `assert` statements cross the whole unit test code into `self.assert*` (as part of deprecating older Pythons). Just for sake of uniformity.

janpipek commented:
Hi, I like the patch, thank you!

What is most important for me (the primary motivation) is that I can parse my "large" files in 2.6 seconds instead of 11. The most optimized non-generator version was about ~2.4-2.5 which is a negligible difference. I don't work with non-blocking streams and cannot bring much enlightenment into this. The decision is yours.

If you are happy with the current approach, I may start with the nit-picking and micro-optimization suggestions (included some, already).

P.S. I agree with the output streams. Here, I'd guess, the solution is trivial because you will just use `write` and will not seek backwards ever (the methods can construct temporary strings though). The main function (`encodeStream`?) will accept an open stream. If you wish, I can write some code once this big pull request is over; or do you want to make it even bigger?

etingof commented:
> What is most important for me (the primary motivation) is that I can parse my "large" files in 2.6 seconds instead of 11. The most optimized non-generator version was about ~2.4-2.5 which is a

So how does the generator version stand performance-wise?

> negligible difference. I don't work with non-blocking streams and cannot bring much enlightenment into this. The decision is yours.

It seems that to support non-blocking streams in addition to blocking streams we need just one `if` statement:

```
received = stream.read(size)
if received is None:
    yield
```

So I guess it's cheap and might be useful for natively non-blocking apps.

> If you are happy with the current approach, I may start with the nit-picking and micro-optimization suggestions (included some, already).

So far I do not expect it to be overhauled again. The things I have in mind are:

* To make the decoder yielding the topmost (or the whole chain?) of ASN.1 objects being populated from the serialization. The reason is to support the use case when incoming stream is too large to fit in memory so the user may want to purge ASN.1 object (e.g. `SequenceOf`) from its head.

* May be slightly unify the `native` encoder with BER decoder. Perhaps just from API standpoint, JSON must be served all at once anyway.

* Add proper docstrings and changelog

So whenever you have time for review, just leave the comments and I will address them.

> P.S. I agree with the output streams. Here, I'd guess, the solution is trivial because you will just use `write` and will not seek backwards ever (the methods can construct temporary strings though). 

Perhaps. Or may be we should wrap it with our seekable stream wrapper to hide the differences among streams and blocking/non-blocking stream behavior?

> The main function (`encodeStream`?) will accept an open stream. 

Note that in the latest edition I tried to reshape it into an iterator.

> If you wish, I can write some code once this big pull request is over; or do you want to make it even bigger?

Please, do if you have time! That would be awesome!

janpipek commented:
> So how does the generator version stand performance-wise?

It stands fine - ~5 % slower than original with some potential for improvement. Perhaps the `read` calls won't allow much but still there are some shortcuts :-)

etingof commented:
> > Is this reuse of instance motivated by performance? Otherwise it looks like a strange pattern.

You are right! This approach might have nasty side-effects. Changed in the latest version of this patch.

etingof commented:
@janpipek I think I am finally done with my insanely disturbing patch! Your nitpicks are welcome!

I propose to merge the entire thing and then tackle the encoders. Then I plan on working on removing Python 2.6 or 2.7 (?) legacy cleaning up the code base a great deal.

janpipek commented on pyasn1/codec/ber/decoder.py:
```diff
@@ -22,165 +21,16 @@
 
 noValue = base.noValue
 
+SubstrateUnderrunError = error.SubstrateUnderrunError
```
Nit pick: `from pyasn1.error import ...` would be more Pythonic.

etingof replied:
Yes, however we have to import a handful of exceptions from that module as well. Should we import each individually? This one is a lookup shortcut...

janpipek commented on pyasn1/codec/ber/decoder.py:
```diff
@@ -5,15 +5,14 @@
 
 from pyasn1 import debug
 from pyasn1 import error
+from pyasn1.codec import streaming
```
```suggestion
from pyasn1.codec.streaming import read, isEndOfStream, asSeekableStream
```
I suggest to import these objects and use them directly (as shown below, every module member lookup adds ~20 ns), basically anywhere where `straming.[...]` is used now.

etingof replied:
Done

janpipek commented on pyasn1/codec/ber/decoder.py:
```diff
@@ -221,51 +78,53 @@ def _createComponent(self, asn1Spec, tagSet, value, **options):
+                    substrate, asn1Spec, tagSet, length,
+                    allowEoo=True, **options):
+                if value is eoo.endOfOctets:
+                    break
```
Looks like you can directly return here and remove the second condition.

etingof replied:
Yeah!

Actually, not all decoders actually expose partially decoded contents. For instance, OctetString and BitString decoders collect received substrate internally up to EOO sentinel. This should be improved.

janpipek commented on pyasn1/codec/streaming.py:
```diff
@@ -0,0 +1,240 @@
+    Yields
+    ------
+    : :py:class:`bytes` or :py:class:`str` or None
+        Returns read data or :py:class:`~pyasn1.error.SubstrateUnderrunError`
```
```suggestion
        Read data or :py:class:`~pyasn1.error.SubstrateUnderrunError`
```

janpipek commented on pyasn1/codec/streaming.py:
```diff
@@ -0,0 +1,240 @@
+
+    Yields
+    ------
+    : :py:class:`bytes` or :py:class:`str` or None
```
```suggestion
    : :py:class:`bytes` or :py:class:`str` or :py:class:`SubstrateUnderrunError`
```

janpipek commented on pyasn1/codec/streaming.py:
```diff
@@ -0,0 +1,240 @@
+        # this will block unless stream is non-blocking
+        received = substrate.read(size)
+        if received is None:  # non-blocking stream can do this
+            yield error.SubstrateUnderrunError(context=context)
```
Philosophical comment: Yielding an error and using it at different places as both a marker and as a real exception is a bit unusual. But I get it.

etingof replied:
Yeah, an alternative would be to return a tuple of (`asn1Object`, `status`) or some other sort of container structure to keep all the details inside... Not sure what's more convenient from user perspective.

WDYT?

etingof replied:
Or may be we could just come up with some dedicated sentinel instead of Exception based `SubstrateUnderrunError`...? Would that be clearer?

janpipek commented on pyasn1/codec/ber/decoder.py:
```diff
@@ -1753,25 +1792,187 @@ def __call__(self, substrate, asn1Spec=None,
+        the input stream, then call the iterator again. The decoder will resume
+        the decoding process using the newly arrived data.
+
+        The `context` property of :py:class:`~pyasn1.error.SubstrateUnderrunError`
```
This is a nice feature.

janpipek commented:
It's nice architecture, well-documented. I'd merge it immediately ;-)

Performance-wise:

![image](https://user-images.githubusercontent.com/92804/65888086-3cbf8380-e39f-11e9-88fa-4f4f0c5eb44a.png)

I see that for small objects, the performance gets a hit. For large files, I am at about 10-20 % loss. This is fine for me. Not sure if that is fine for you.

janpipek reviewed (commented):
A few comments. Perhaps not complete - perhaps the most important is the one about performance.

etingof commented:
> It's nice architecture, well-documented. I'd merge it immediately ;-)
> 
> Performance-wise:
> 
> ![image](https://user-images.githubusercontent.com/92804/65888086-3cbf8380-e39f-11e9-88fa-4f4f0c5eb44a.png)
> 
> I see that for small objects, the performance gets a hit. For large files, I am at about 10-20 % loss. This is fine for me. Not sure if that is fine for you.

That's too slow indeed... Can you see the hot spots in your profiling? Anything obvious to optimize, apart from those places you've already mentioned?

janpipek commented:
> That's too slow indeed... Can you see the hot spots in your profiling? Anything obvious to optimize, apart from those places you've already mentioned?

I'll try tomorrow. Now I have to leave :-( The places I mentioned should not be responsible for the gap.

etingof committed:
Update docstring on streaming.read

Co-Authored-By: Jan Pipek <jan.pipek@gmail.com>

etingof committed:
Update docstring on streaming.read

Co-Authored-By: Jan Pipek <jan.pipek@gmail.com>

etingof committed:
Update `streaming.read` docstring

Co-Authored-By: Jan Pipek <jan.pipek@gmail.com>

etingof committed:
Update `streaming.read` docstring

Co-Authored-By: Jan Pipek <jan.pipek@gmail.com>

etingof committed:
Optimize `streaming` objects access for performance

etingof committed:
Optimize `streaming` objects access for performance

etingof commented:
> It's nice architecture, well-documented. I'd merge it immediately ;-)
> 
> Performance-wise:
> 
> ![image](https://user-images.githubusercontent.com/92804/65888086-3cbf8380-e39f-11e9-88fa-4f4f0c5eb44a.png)
> 
> I see that for small objects, the performance gets a hit. For large files, I am at about 10-20 % loss. This is fine for me. Not sure if that is fine for you.

I can't find your benchmark for the unmodified (master) code. I hope with my latest hacks we are not back to the original slowness...? ;-)

etingof commented on pyasn1/codec/ber/decoder.py:
```diff
@@ -1698,25 +1732,30 @@ def __call__(self, substrate, asn1Spec=None,
+                            yield value
+
+                    bytesRead = substrate.tell() - original_position
+                    if bytesRead != length:
```
I wonder if we need this check at all? Why don't we trust `valueDecoder` to do the right thing?

janpipek replied:
This does not seem to have any measurable performance effect. I thought this would be a nice cross-check but you can drop it if you trust the called functions.

etingof commented on pyasn1/codec/ber/decoder.py:
```diff
@@ -276,24 +135,27 @@ def valueDecoder(self, substrate, asn1Spec,
-        if not the_bytes:
-            return self._createComponent(asn1Spec, tagSet, 0, **options)
+        for chunk in streaming.read(substrate, length, options):
+            if isinstance(chunk, SubstrateUnderrunError):
```
I am still suspicious wrt to `isinstance` versus `is` in all these conditions. Would be interesting to replace one with the other and benchmark the difference...

etingof commented on pyasn1/codec/ber/decoder.py:
```diff
@@ -1753,25 +1792,187 @@ def __call__(self, substrate, asn1Spec=None,
+        self._substrate = streaming.asSeekableStream(substrate)
+        self._asn1Spec = asn1Spec
+        self._options = kwargs
+        self._decoder = self.SINGLE_ITEM_DECODER()
```
This is my other suspect that can cause the slowdown in the recent benchmarks. This object collects and maintains a cache of decoded BER tags. Previously, that cache never expired and might be heavily reused. Now it is only used once per `StreamingDecoder` call. On top of that, cache maintenance for short objects might actually impose performance penalty rather than benefits.

Would be interesting to temporary make this object global and thus reusable to see what's the difference.

etingof committed:
Add minor performance optimising changes

etingof committed:
Add minor performance optimising changes

janpipek commented:
> I can't find your benchmark for the unmodified (master) code. I hope with my latest hacks we are not back to the original slowness...? ;-)

Unfortunately, "old code" here means the unmodified master. I actually never got much below its performance. The optimizing changes sutbract ~2 us from all three values.

etingof commented:
> Unfortunately, "old code" here means the unmodified master. I actually never got much below its performance. The optimizing changes sutbract ~2 us from all three values.

My impression is that for smaller objects, you original streaming design was actually a tad bit slower. However, for larger objects the difference in performance was whooping (you mentioned minutes vs an hour)!

janpipek commented:
Hi, I transformed the benchmarking notebook (for small objects) into an automatic script:

https://gist.github.com/janpipek/4d0c73341f6457c9770ec5804d62bf45

- As you can see, for "real" data (rfc2314), the support for suspendible generator is ~20 % slower.

- For a 2 MB file (our proprietary), it's also ~20 % slower than master.

- For the huge example you mentioned, the result is: master ~1hr, streams without yield ~1:36 min, streams with yield ~1:52 min (hard to be precise on a CPU with automatic modulation :-( ). There is a measurable difference but not a huge one.

etingof commented:
Thank you! Let me play with this, may be I could optimize something else...

How do you count performance difference of 20%? It seems that the yielding version takes 17s where other versions take 7-8s. Which is over 100%...

janpipek commented:
From the examples, the last one is a certificate from pyasn1_examples. That takes ~20 % more.

etingof commented:
I am referring to the first test (false) - it seems that my version is two times (100%) slower than master, no?

```
master
Switched to branch 'master'
Your branch is up to date with 'origin/master'.
false
7.23 ± 0.12 us
example1
33.18 ± 0.31 us
asn
53.66 ± 0.28 us
pem
616.09 ± 4.11 us
support-incomplete-substrate
Switched to branch 'support-incomplete-substrate'
Your branch is up to date with 'origin/support-incomplete-substrate'.
false
17.0 ± 0.16 us
example1
60.35 ± 0.24 us
asn
90.99 ± 0.25 us
pem
750.14 ± 2.04 us
```

janpipek commented:
Where do the penalties come from? Just some hypothesising...

- To start with, every "yield" on its own takes ~100 ns:

![image](https://user-images.githubusercontent.com/92804/66050843-37dd0a00-e52e-11e9-87f5-193dd7ebeb31.png)

- If you wrap the read for an already reasonable stream in a separate method (+ the necessary yield mechanism), you pay ~350-400 ns per read:

![image](https://user-images.githubusercontent.com/92804/66051456-41b33d00-e52f-11e9-8985-dbe5b97336eb.png)

janpipek commented:
> I am referring to the first test (false) - it seems that my version is two times (100%) slower than master, no?

Yes, that is true. However, it depends on what is your aim. It seems natural that the more general construction you build around reading three bytes from a string, the less effective it becomes (as extreme example, you can just check whether the string matches the tree bytes for True / False to obtain sub-microsecond speed ;-)).

The price for indirection is less pronounced when large structures are parsed, which I think is what matters. Based on the experimentation with yields etc., I am only a bit sceptical about reducing the 20 % gap for larger objects to zero because the incomplete stream support is adding complexity and checks "by definition". The question is how small this gap can be.

As far as I am concerned, I do not need incomplete streams. Nor do I care too much about the 20 %. It's not a big problem to have 20 % more computing power, it's a problem to have 4x or 30x times the computing power.

I'll keep looking for places to optimize.

etingof commented:
Right, I just wanted to make sure that our systems of measurement are the same. ;-)

> As far as I am concerned, I do not need incomplete streams.

My main motivation here is the use-case of a streaming network connection. BER is heavily used in telecom and networking protocols (AFAIK, the entire GSM/3G/4G/5G stack is based on ASN.1/BER). So it seemingly makes sense to be able to suspend the decoder (rather than track back and restart) if input data is not the complete serialization...

etingof commented:
@janpipek JFYI: I am using [this](https://gist.github.com/etingof/a26f43f2f0a1ce82d675b99fba2bd8b9) simple profiling script based on your benchmarking script in hope to find out the cause of the slowdown.

janpipek commented:
> My main motivation here is the use-case of a streaming network connection...

Well, then the applicability comes first and performance second. I understand :-)

P.S. Our usage is also in analysing telecom data, although never real-time.

etingof committed:
Reuse `SingleItemDecoder` object in `StreamingDecoder`

Try to reuse `SingleItemDecoder` object to leverage its caches.

etingof committed:
Reuse `SingleItemDecoder` object in `StreamingDecoder`

Try to reuse `SingleItemDecoder` object to leverage its caches.

etingof commented:
Commit 0261649 gives us ~15% back on smallish objects by reusing tag cache.

etingof commented:
> P.S. Our usage is also in analysing telecom data, although never real-time.

Interestingly, back then they used ASN.1 for everything e.g. real-time signaling, media streaming, usage accounting records etc.

etingof commented:
Globally replacing code like this:

```
if isinstance(component, SubstrateUnderrunError):
    yield component
```

with code like this:

```
if component is substrateUnderrunError:
    yield component
```

Improves performance by 8% on smaller and 6% on larger objects.... Any suggestions? %-/

janpipek commented:
This is optimistic :-)

janpipek commented:
Sorry, I got distracted and now I am bit lost in the status of this PR - should I merge your last version into my branch and then you my branch into `master`? Or do you think we should aim for further optimizations? I'd be happy to take any path :-)

etingof commented:
> Sorry, I got distracted and now I am bit lost in the status of this PR - should I merge your last version into my branch and then you my branch into `master`? Or do you think we should aim for further optimizations? I'd be happy to take any path :-)

Do did I! But I am now in position to focus on finalizing this PR.

etingof committed:
Allow combining constraints operating on sets (#181)

Added ability of combining `SingleValueConstraint` and
`PermittedAlphabetConstraint` objects into one for proper
modeling `FROM ... EXCEPT ...` ASN.1 clause.

etingof committed:
Fix Integer decoder to handle empty payload

etingof committed:
Fix Integer decoder to handle empty payload

etingof committed:
Merge branch 'streamoptimum37' of github.com:janpipek/pyasn1 into streamoptimum37

etingof committed:
Merge branch 'support-incomplete-substrate' of github.com:janpipek/pyasn1 into support-incomplete-substrate

etingof committed:
Fix streaming unit test

etingof committed:
Refactor codec classes linkage

Make it looking more uniform and easier to override if needed.

etingof commented:
Hey @janpipek o/

I think this change is reasonably complete feature-wise by now.  I still have some refactoring ideas, but those can follow up.

Next thing we need to do is to deprecate Py2.5 and may be 2.6 support for this new code to work. This is going to be a follow up as well, otherwise it'd be overly messy.

Finally, `yield from` looks way better than those silly loops.... The other option would be to branch pyasn1 into Py3.4+ version. That would so much simplify the codebase that it's almost irresistible to undertake. Then we'd end up with pyasn1-0.5+ supporting py2.6+ and pyasn1-1.x supporting Py3.4+.

Thoughts?

etingof committed:
Pass `tagMap` and `typeMap` to decoder instance

This change should simplify decoder specialization by means
of parameterization in addition to subclassing.

etingof commented:
I think we can push this further and address whatever concerns in follow-ups.
